ruby/red-arrow/lib/arrow/decimal128-array.rb

```diff
@@ -18,7 +18,9 @@
 module Arrow
   class Decimal128Array
     def get_value(i)
-      BigDecimal(format_value(i))
+      string = format_value(i)
+      string.sub!(".E", ".0E") if string.include?(".E")
```

Comment:
If I may ask, isn't this quite inefficient, having to scan the string multiple times like this?  

Would it be more efficient to process the high bits and low bits of the Arrow Decimal instead?

Reply:
Hmm. Could you provide a code for your suggested approach?

FYI: Here is a benchmark script:

```ruby
#!/usr/bin/env ruby

require "benchmark"

n = 100000
Benchmark.bmbm do |x|
  x.report("0.0E - nothing") do
    n.times do
      string = "0.0E"
      string
    end
  end

  x.report("0.E - nothing") do
    n.times do
      string = "0.E"
      string
    end
  end

  x.report("0.0E - include?") do
    n.times do
      string = "0.0E"
      string.sub!(".E", ".0E") if string.include?(".E")
      string
    end
  end

  x.report("0.E - include?") do
    n.times do
      string = "0.E"
      string.sub!(".E", ".0E") if string.include?(".E")
      string
    end
  end
end
```

```text
Rehearsal ---------------------------------------------------
0.0E - nothing    0.007074   0.000452   0.007526 (  0.007524)
0.E - nothing     0.006988   0.000000   0.006988 (  0.006988)
0.0E - include?   0.015868   0.000000   0.015868 (  0.015869)
0.E - include?    0.066138   0.000000   0.066138 (  0.066140)
------------------------------------------ total: 0.096520sec

                      user     system      total        real
0.0E - nothing    0.006819   0.000000   0.006819 (  0.006819)
0.E - nothing     0.006878   0.000000   0.006878 (  0.006878)
0.0E - include?   0.015696   0.000000   0.015696 (  0.015697)
0.E - include?    0.066329   0.000000   0.066329 (  0.066330)
```

Reply:
ah sorry for not being clear.  
I was thinking in the context of Arrow IPC, where the decimal data is already [encoded using 2 int64](https://github.com/apache/arrow/blob/main/go/arrow/decimal128/decimal128.go#L49) (sorry I'm still looking for the C++ implementation).

Though perhaps this code isn't about converting Arrow data into Ruby?

Reply:
Correction: It seems Arrow encodes decimal as fixed-size byte array. The two-complement representation is more of an implementation/runtime thing.  
While Ruby BigDecimal cannot be constructed from a byte array anyway (?)

Reply:
Arrow's decimal is fixed size value but BigDeciaml isn't fixed size value.
BigDecimal doesn't provide an API that constructs a `BigDecimal` from binary data.
So we need to use `String` not binary data to constructs a `BigDecimal` without data loss.

Reply:
I see. Thank you for explaining to me.

Would `string.sub!(".E", ".0E")` (without `include?`) still be better?  
It saves 1 string scan and is shown to be faster in the above benchmark.

Reply:
It's slow for non-`0.E` cases. I think that `0.E` case is a rare case. So I think that we should prioritize non-`0.E` cases than `0.E` case.

```ruby
#!/usr/bin/env ruby

require "benchmark"

n = 100000
Benchmark.bmbm do |x|
  x.report("0.0E - nothing") do
    n.times do
      string = "0.0E"
      string
    end
  end

  x.report("0.E - nothing") do
    n.times do
      string = "0.E"
      string
    end
  end

  x.report("0.0E - include?") do
    n.times do
      string = "0.0E"
      string.sub!(".E", ".0E") if string.include?(".E")
      string
    end
  end

  x.report("0.E - include?") do
    n.times do
      string = "0.E"
      string.sub!(".E", ".0E") if string.include?(".E")
      string
    end
  end

  x.report("0.0E - always sub!") do
    n.times do
      string = "0.0E"
      string.sub!(".E", ".0E")
      string
    end
  end

  x.report("0.E - always sub!") do
    n.times do
      string = "0.E"
      string.sub!(".E", ".0E")
      string
    end
  end
end
``` 

```text
Rehearsal ------------------------------------------------------
0.0E - nothing       0.007245   0.000000   0.007245 (  0.007261)
0.E - nothing        0.006175   0.000441   0.006616 (  0.006616)
0.0E - include?      0.015160   0.000000   0.015160 (  0.015161)
0.E - include?       0.064167   0.000000   0.064167 (  0.064181)
0.0E - always sub!   0.028071   0.000000   0.028071 (  0.028070)
0.E - always sub!    0.054682   0.000000   0.054682 (  0.054683)
--------------------------------------------- total: 0.175941sec

                         user     system      total        real
0.0E - nothing       0.006584   0.000000   0.006584 (  0.006584)
0.E - nothing        0.006586   0.000000   0.006586 (  0.006585)
0.0E - include?      0.014944   0.000000   0.014944 (  0.014944)
0.E - include?       0.063128   0.000000   0.063128 (  0.063129)
0.0E - always sub!   0.027945   0.000000   0.027945 (  0.027946)
0.E - always sub!    0.054426   0.000000   0.054426 (  0.054426)
```